File: Stage22/C/Problem25682.c
#include <stdio.h>

int N, M, K;
char board[2001][2001];
int psumW[2001][2001];
int psumB[2001][2001];

int min(int a, int b) {
    return a < b ? a : b;
}

int main(void) {
    scanf("%d %d %d", &N, &M, &K);

    for (int i = 1; i <= N; i++) {
        scanf("%s", board[i] + 1);
    }

    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= M; j++) {
            int isWhite = ((i + j) % 2 == 0);

            int w = (board[i][j] == (isWhite ? 'W' : 'B')) ? 0 : 1;
            int b = (board[i][j] == (isWhite ? 'B' : 'W')) ? 0 : 1;

            psumW[i][j] = psumW[i - 1][j] + psumW[i][j - 1]
                        - psumW[i - 1][j - 1] + w;

            psumB[i][j] = psumB[i - 1][j] + psumB[i][j - 1]
                        - psumB[i - 1][j - 1] + b;
        }
    }

    int ans = 1e9;

    for (int i = K; i <= N; i++) {
        for (int j = K; j <= M; j++) {
            int w = psumW[i][j] - psumW[i - K][j]
                  - psumW[i][j - K] + psumW[i - K][j - K];

            int b = psumB[i][j] - psumB[i - K][j]
                  - psumB[i][j - K] + psumB[i - K][j - K];

            ans = min(ans, min(w, b));
        }
    }

    printf("%d\n", ans);
    return 0;
}

File: Stage22/Java/Problem25682.java
package Stage22.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Problem25682 {
    static int N, M, K;
    static char[][] board;
    static int[][] psumW, psumB;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        board = new char[N + 1][M + 1];
        psumW = new int[N + 1][M + 1];
        psumB = new int[N + 1][M + 1];

        for (int i = 1; i <= N; i++) {
            String line = br.readLine();
            for (int j = 1; j <= M; j++) {
                board[i][j] = line.charAt(j - 1);
            }
        }

        for (int i = 1; i <= N; i++) {
            for (int j = 1; j <= M; j++) {
                boolean isWhite = (i + j) % 2 == 0;

                int w = (board[i][j] == (isWhite ? 'W' : 'B')) ? 0 : 1;
                int b = (board[i][j] == (isWhite ? 'B' : 'W')) ? 0 : 1;

                psumW[i][j] = psumW[i - 1][j] + psumW[i][j - 1]
                            - psumW[i - 1][j - 1] + w;

                psumB[i][j] = psumB[i - 1][j] + psumB[i][j - 1]
                            - psumB[i - 1][j - 1] + b;
            }
        }

        int ans = Integer.MAX_VALUE;

        for (int i = K; i <= N; i++) {
            for (int j = K; j <= M; j++) {
                int w = psumW[i][j] - psumW[i - K][j]
                      - psumW[i][j - K] + psumW[i - K][j - K];

                int b = psumB[i][j] - psumB[i - K][j]
                      - psumB[i][j - K] + psumB[i - K][j - K];

                ans = Math.min(ans, Math.min(w, b));
            }
        }

        System.out.println(ans);
        br.close();
    }
}
